crypto/mem_sec.c

```diff
@@ -381,17 +381,33 @@ static int sh_init(size_t size, size_t minsize)
 
     memset(&sh, 0, sizeof(sh));
 
-    /* make sure size and minsize are powers of 2 */
+    /* make sure size is a powers of 2 */
     OPENSSL_assert(size > 0);
     OPENSSL_assert((size & (size - 1)) == 0);
-    OPENSSL_assert((minsize & (minsize - 1)) == 0);
     if (size <= 0 || (size & (size - 1)) != 0)
         goto err;
-    if (minsize == 0 || (minsize & (minsize - 1)) != 0)
-        goto err;
 
-    while (minsize < (int)sizeof(SH_LIST))
-        minsize *= 2;
+    if (minsize <= sizeof(SH_LIST)) {
+        OPENSSL_assert(sizeof(SH_LIST) <= 65536);
+        /*
+         * Compute the minimum possible allocation size.
+         * This must be a power of 2 and at least as large as the SH_LIST
+         * structure.
+         */
+        minsize = sizeof(SH_LIST) - 1;
+        minsize |= minsize >> 1;
+        minsize |= minsize >> 2;
+        if (sizeof(SH_LIST) > 16)
```

Comment:
Actually, one thought.  Branches are much more expensive than a `shiff` and `or`.  Just remove the conditionals and always shift by 4 and 8.

Reply:
Any compiler that generates a branch here is really bad at optimising.  The expression is a constant.  The idea is to allow a compiler to completely remove the shift/or.  The structure contains two pointers, I very much doubt that this will be more than 16 bytes on any architecture we care about now or for a long time into the future.

Reply:
Fair enough, yes the condition is constant.